### PR TITLE
WIP: Add Applicability to suggestion lints

### LINKED ISSUE #2930: Use suggestion applicabilities everywhere
rust lets you [tag suggestions with an applicability level](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Diagnostic.html#method.span_suggestion_with_applicability), so that tools like rustfix know which suggestions are applicable.

We should use this here too. `span_to_snippet` calls should generate `HasPlaceholders` applicabilities during the fallback, and if macros are involved we should use MaybeIncorrect, perhaps (or add a InMacro applicability).

Currently I don't want to advertise rustfix too much since the UI is edition focused, but once that settles we should be prepared to have rustfix work well with clippy.

### PR DESCRIPTION
**Pull request by flip1995**
Resolves #2930

The applicability levels added to the `span_lint_and_sugg` and `multispan_sugg` (do we need this on `multispan_sugg`?) functions are how I felt would be the best fitting ones.

I haven't tested this yet, because I wanted to wait for the new nightly instead of doing this manual. :smile: 

cc @Manishearth

### CODE REVIEW

File: clippy_lints/src/bytecount.rs [hunk 1]
```diff
@@ -86,7 +87,8 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ByteCount {
                                            "Consider using the bytecount crate",
                                            format!("bytecount::count({}, {})",
                                                     snippet(cx, haystack.span, ".."),
-                                                    snippet(cx, needle.span, "..")));
+                                                    snippet(cx, needle.span, "..")),
+                                           Applicability::HasPlaceholders);
```
Comment (oli-obk) [inline clippy_lints/src/bytecount.rs]: It only "maybe" has placeholders. We should make the `snippet` function return a tuple of `Applicability` and snippet-`String`

**Comment by Manishearth** [comment]
FWIW http://github.com/kennytm/rustup-toolchain-install-master lets you locally build clippy on trunk

File: clippy_lints/src/bytecount.rs (hunk 1 above)
Comment (Manishearth) [reply clippy_lints/src/bytecount.rs]: The best thing to do here would be to have it accept `&mut Applicability` IMO

**Comment by flip1995** [comment]
> FWIW http://github.com/kennytm/rustup-toolchain-install-master lets you locally build clippy on trunk

Nice to know! Thanks!

**Comment by killercup** [comment]
Thanks so much for doing this!

**Comment by oli-obk** [comment]
needs a rebase

**Comment by flip1995** [comment]
Should I add the Applicability argument to the snippet function in this PR or in a separate one?

File: clippy_lints/src/methods.rs [hunk 2]
```diff
@@ -1325,6 +1332,7 @@ fn lint_unnecessary_fold(cx: &LateContext<'_, '_>, expr: &hir::Expr, fold_args:
                     "this `.fold` can be written more succinctly using another method",
                     "try",
                     sugg,
+                    Applicability::HasPlaceholders,
```
Comment (shssoichiro) [inline clippy_lints/src/methods.rs]: It may be good to change this to `Applicability::MaybeIncorrect`. See #3069 and #3303.

Comment (flip1995) [reply clippy_lints/src/methods.rs]: Oh this is a stale PR to remind me (and others) that this should be done sometimes. When I'm (or someone else) continues working on this, I would start from scratch and go over these `Applicability` levels again (Also taking #3191 into consideration).

But yeah `Applicability::MaybeIncorrect` would probably be better here.

**Comment by flip1995** [comment]
Closing in favor of #3459

**Pull request closed without merging**